torch/nn/modules/transformer.py

```diff
@@ -17,13 +17,17 @@
 
 def _generate_square_subsequent_mask(
         sz: int,
-        device: torch.device = torch.device(torch._C._get_default_device()),  # torch.device('cpu'),
-        dtype: torch.dtype = torch.get_default_dtype(),
+        device: Optional[torch.device] = None,
+        dtype: Optional[torch.dtype] = None,
 ) -> Tensor:
     r"""Generate a square causal mask for the sequence.
 
     The masked positions are filled with float('-inf'). Unmasked positions are filled with float(0.0).
     """
+    if device is None:
+        device = torch.device('cpu')
```

Comment:
Why aren't we using torch._C._get_default_device() anymore? Couldn't that also return cuda?

Reply:
@soulitzer sorry I was slow in commenting, see https://github.com/pytorch/pytorch/pull/115867#discussion_r1427290212

Reply:
Oh what I mean is to do something like:
```
def _generate_square_subsequent_mask(
        sz: int,
        device: Optional[torch.device] = None,
        dtype: Optional[torch.dtype] = None,
) -> Tensor:
    if device is None:
        device = torch._C._get_default_device()
   [...]
```
Does that work?

Reply:
My bad, I had thought that  `torch.set_default_device('cuda')` would change the type returned by `torch._C._get_default_device()` so I thought that the device of the tensor returned by `_generate_square_subsequent_mask` didn't change regardless of what I was setting the device to.

If I understand correctly seems that the only way to set the device gotten by `torch._C._get_default_device` is via `torch.set_default_tensor_type(torch.cuda.FloatTensor)` which is deprecated as of 2.1  😛  

Nevertheless changed these to still call the functions

Reply:
If only there were a way to get the status of whatever `torch.set_default_device` or `with torch.device("cuda")` sets though 🤔  
I'm also not sure what is `torch.set_default_tensor_type(torch.cuda.FloatTensor)` is supposed to do...
```
torch.set_default_tensor_type(torch.cuda.FloatTensor)
 print(torch._C._get_default_device())  # cuda
 a = torch.zeros(1)   # cpu
 ```